missing digest on `main` and `2.13.x` branch

Observed a missing digest on `main` even with backtracking changes in place:

```
Engine traceback:
  in select
  in pants.core.goals.test.run_tests
  in pants.backend.python.goals.pytest_runner.run_python_test (src/python/pants/jvm/compile_test.py:tests)
  in pants.backend.python.goals.pytest_runner.setup_pytest_for_target
  in pants.backend.python.util_rules.pex.create_venv_pex (pytest_runner.pex)
  in pants.backend.python.util_rules.pex.build_pex (pytest_runner.pex)
  in pants.engine.process.fallible_to_exec_result_or_raise
  in process
Traceback (no traceback):
  <pants native internals>
Exception: Could not recover from missing digest: Was not present in either the local or remote store: Digest { hash: Fingerprint<217f1d5845a05a89384e5d1e58b23372731b447685ef36c4b204845c2d6ae231>, size_bytes: 2768 }
Error: Process completed with exit code 1.
```

https://github.com/pantsbuild/pants/runs/7113601869?check_suite_focus=true#step:13:424

Another example: 

```
Engine traceback:
  in select
  in pants.core.goals.lint.lint
  in pants.backend.scala.lint.scalafmt.rules.scalafmt_fmt (scalafmt)
  in pants.engine.process.fallible_to_exec_result_or_raise
  in process
Traceback (no traceback):
  <pants native internals>
Exception: Could not recover from missing digest: Was not present in either the local or remote store: Digest { hash: Fingerprint<9484ac95856510459d1bd52a77a6b93cdd641560decdf9910395ee4d17e88163>, size_bytes: 121417 }
Error: Process completed with exit code 1.
```

https://github.com/pantsbuild/pants/runs/7116113010?check_suite_focus=true#step:9:279

Same build failed again but only with local store this time: 

```
Engine traceback:
  in select
  in pants.core.goals.lint.lint
  in pants.backend.python.lint.black.rules.black_fmt (black)
  in pants.engine.process.fallible_to_exec_result_or_raise
  in process
Traceback (no traceback):
  <pants native internals>
Exception: Could not recover from missing digest: Was not present in the local store: Digest { hash: Fingerprint<740528d33c19eca061f11cbee7fd7b030928a8b7b53a5023da6e6168236cc0b9>, size_bytes: 1817 }
Error: Process completed with exit code 1.
```

https://github.com/pantsbuild/pants/runs/7116545174?check_suite_focus=true#step:9:242

--- a/pants.toml
+++ b/pants.toml
@@ -68,11 +68,8 @@ unmatched_build_file_globs = "error"
 # NB: Users must still set `--remote-cache-{read,write}` to enable the remote cache.
 remote_store_address = "grpcs://cache.toolchain.com:443"
 remote_auth_plugin = "toolchain.pants.auth.plugin:toolchain_auth_plugin"
-
 # See https://github.com/pantsbuild/pants/issues/11331.
-# N.B.: Switched back to eager due to issues documented in:
-#   https://github.com/pantsbuild/pants/issues/15995
-remote_cache_eager_fetch = true
+remote_cache_eager_fetch = false
 
 [anonymous-telemetry]
 enabled = true

--- a/src/rust/engine/src/context.rs
+++ b/src/rust/engine/src/context.rs
@@ -22,6 +22,7 @@ use async_oncecell::OnceCell;
 use cache::PersistentCache;
 use fs::{safe_create_dir_all_ioerror, GitignoreStyleExcludes, PosixFS};
 use graph::{self, EntryId, Graph, InvalidationResult, NodeContext};
+use hashing::Digest;
 use log::info;
 use parking_lot::Mutex;
 use process_execution::{
@@ -650,10 +651,12 @@ pub struct Context {
   run_id: RunId,
   /// The number of attempts which have been made to backtrack to a particular ExecuteProcess node.
   ///
-  /// Presence in this map at process runtime indicates that the pricess is being retried, and that
+  /// Presence in this map at process runtime indicates that the process is being retried, and that
   /// there was something invalid or unusable about previous attempts. Successive attempts should
   /// run in a different mode (skipping caches, etc) to attempt to produce a valid result.
   backtrack_levels: Arc<Mutex<HashMap<ExecuteProcess, usize>>>,
+  /// The Digests that we have successfully invalidated a Node for.
+  backtrack_digests: Arc<Mutex<HashSet<Digest>>>,
   stats: Arc<Mutex<graph::Stats>>,
 }
 
@@ -666,6 +669,7 @@ impl Context {
       session,
       run_id,
       backtrack_levels: Arc::default(),
+      backtrack_digests: Arc::default(),
       stats: Arc::default(),
     }
   }
@@ -704,7 +708,7 @@ impl Context {
       return result;
     };
 
-    // Locate the source(s) of this Digest and their backtracking levels.
+    // Locate live source(s) of this Digest and their backtracking levels.
     // TODO: Currently needs a combination of `visit_live` and `invalidate_from_roots` because
     // `invalidate_from_roots` cannot view `Node` results. Would be more efficient as a merged
     // method.
@@ -719,12 +723,23 @@ impl Context {
     });
 
     if candidate_roots.is_empty() {
-      // We did not identify any roots to invalidate: allow the Node to fail.
-      return result.map_err(|e| {
-        throw(format!(
-          "Could not identify a process to backtrack to for: {e}"
-        ))
-      });
+      // If there are no live sources of the Digest, see whether any have already been invalidated
+      // by other consumers.
+      if self.backtrack_digests.lock().get(&digest).is_some() {
+        // Some other consumer has already identified and invalidated the source of this Digest: we
+        // can wait for the attempt to complete.
+        return Err(Failure::Invalidated);
+      } else {
+        // There are no live or invalidated sources of this Digest. Directly fail.
+        return result.map_err(|e| {
+          throw(format!(
+            "Could not identify a process to backtrack to for: {e}"
+          ))
+        });
+      }
+    } else {
+      // We have identified a Node to backtrack on. Record it.
+      self.backtrack_digests.lock().insert(digest);
     }
 
     // Attempt to trigger backtrack attempts for the matched Nodes. It's possible that we are not
@@ -807,6 +822,7 @@ impl NodeContext for Context {
       session: self.session.clone(),
       run_id: self.run_id,
       backtrack_levels: self.backtrack_levels.clone(),
+      backtrack_digests: self.backtrack_digests.clone(),
       stats: self.stats.clone(),
     }
   }

